File: python-ocr-service/app/mrz_parser.py
"""
FastMRZ Parser for ICAO 9303 Machine Readable Zone

This module uses FastMRZ to parse and validate MRZ text extracted by OCR.
"""
import logging
import sys
from typing import Optional, Dict, Any
from datetime import datetime
from fastmrz import FastMRZ

# Configure logger to output to stderr (always visible)
logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)
if not logger.handlers:
    handler = logging.StreamHandler(sys.stderr)
    handler.setLevel(logging.INFO)
    formatter = logging.Formatter('[%(asctime)s] [MRZ_PARSER] %(levelname)s - %(message)s')
    handler.setFormatter(formatter)
    logger.addHandler(handler)


class MRZParser:
    """
    FastMRZ wrapper for parsing and validating passport MRZ data.

    Features:
    - ICAO 9303 compliant parsing
    - Check digit validation
    - Date format conversion
    - Error correction for common OCR mistakes
    """

    def __init__(self):
        """Initialize MRZ parser"""
        # FastMRZ is a function-based parser, not a class
        # We'll parse MRZ manually using ICAO 9303 format
        logger.info("MRZ parser initialized")

    def parse(self, mrz_text: str) -> Optional[Dict[str, Any]]:
        """
        Parse MRZ text into structured passport data.

        Args:
            mrz_text: 88-character MRZ string (2 lines × 44 chars)

        Returns:
            Dictionary with parsed fields or None if invalid:
            {
                'passportNumber': str,
                'surname': str,
                'givenName': str,
                'nationality': str,
                'dateOfBirth': str (YYYY-MM-DD),
                'sex': str ('M' or 'F'),
                'dateOfExpiry': str (YYYY-MM-DD),
                'issuingCountry': str,
                'personalNumber': str (optional),
                'validCheckDigits': bool,
                'rawMrz': str
            }

        Example:
            >>> parser = MRZParser()
            >>> data = parser.parse("P<USASMITH<<JOHN<ROBERT<<<...")
            >>> print(data['passportNumber'])
            'N1234567'
        """
        if not mrz_text or len(mrz_text) != 88:
            logger.warning(f"Invalid MRZ length: {len(mrz_text) if mrz_text else 0} (expected 88)")
            return None

        try:
            # Apply OCR error corrections before parsing
            corrected_mrz = self._correct_ocr_errors(mrz_text)

            # Parse MRZ manually (ICAO 9303 format)
            parsed_data = self._extract_fields(None, corrected_mrz)

            if not parsed_data or not parsed_data.get('passportNumber'):
                logger.warning("MRZ parsing returned no passport number")
                return None

            # Validate check digits
            is_valid = self._validate_check_digits_manual(corrected_mrz)
            parsed_data['validCheckDigits'] = is_valid

            if not is_valid:
                logger.warning("MRZ check digit validation failed (but returning data anyway)")

            logger.info(f"Successfully parsed MRZ for passport: {parsed_data.get('passportNumber', 'UNKNOWN')}")
            return parsed_data

        except Exception as e:
            logger.error(f"MRZ parsing failed: {str(e)}")
            return None

    def _correct_ocr_errors(self, mrz_text: str) -> str:
        """
        Apply minimal OCR error corrections.

        CRITICAL: Don't over-correct! PaddleOCR is usually accurate.
        Only fix obvious errors that don't destroy data.

        Strategy: Minimal corrections, preserve original text
        """
        corrected = mrz_text.upper()

        # Only replace spaces with field separators
        corrected = corrected.replace(" ", "<")

        # That's it! Let PaddleOCR's raw output through.
        # The original aggressive corrections were destroying passport numbers.

        return corrected

    def _extract_fields(self, result: Any, raw_mrz: str) -> Dict[str, Any]:
        """
        Extract fields from FastMRZ result object.

        FastMRZ returns object with attributes, convert to dict.
        """
        try:
            # Get line 1 and line 2 from raw MRZ
            line1 = raw_mrz[:44]
            line2 = raw_mrz[44:88]

            # LOG EXACTLY WHAT WE'RE PARSING
            logger.info(f"=== PARSING MRZ ===")
            logger.info(f"Line 1 (44 chars): {line1}")
            logger.info(f"Line 2 (44 chars): {line2}")

            # Parse issuing country from line 1 (positions 2-4)
            issuing_country = line1[2:5].replace("<", "")
            logger.info(f"Issuing Country (pos 2-5): '{issuing_country}'")

            # Parse names from line 1 (positions 5-43)
            # Format: SURNAME<<GIVENNAME1<GIVENNAME2<...
            # But OCR sometimes reads single < instead of <<
            names_raw = line1[5:44]
            logger.info(f"Names raw (pos 5-44): '{names_raw}'")

            # Find the double separator << which separates surname from given names
            if '<<' in names_raw:
                # Standard format with << separator
                parts = names_raw.split('<<', 1)  # Split only on first <<
                surname = parts[0].replace('<', ' ').strip()
                given_name_raw = parts[1].replace('<', ' ').strip() if len(parts) > 1 else ""

                logger.info(f"Found << separator at position {names_raw.index('<<')}")

                # HEURISTIC FIX 1: If given_name is empty but surname is very long (>15 chars),
                # OCR probably merged given names into surname without separator
                # Example: 'NIKOLOV<NIKOLAYSTOYANOV' instead of 'NIKOLOV<NIKOLAY<STOYANOV'
                if not given_name_raw and len(parts[0].replace('<', '')) > 15:
                    # Split the "surname" by <, first part is real surname
                    name_parts = parts[0].split('<')
                    name_parts = [p for p in name_parts if p]  # Remove empty

                    if len(name_parts) >= 2:
                        surname = name_parts[0]
                        # Rest is given name (but merged without separators)
                        given_name_merged = ''.join(name_parts[1:])

                        # Try to split merged given names intelligently
                        # Look for capital letter patterns or common name patterns
                        # For now, just split in half if > 12 chars
                        if len(given_name_merged) > 12:
                            # Likely two names: split in middle
                            mid = len(given_name_merged) // 2
                            given_name = given_name_merged[:mid] + ' ' + given_name_merged[mid:]
                        else:
                            given_name = given_name_merged

                        logger.info(f"Heuristic split applied (empty given name): surname={surname}, merged_given={given_name_merged}, split_given={given_name}")
                    else:
                        given_name = given_name_raw
                # HEURISTIC FIX 2: If given_name exists but has no spaces and is very long (>12 chars),
                # OCR probably merged multiple given names together
                # Example: 'NIKOLAYSTOYANOV' instead of 'NIKOLAY STOYANOV'
                elif given_name_raw and ' ' not in given_name_raw and len(given_name_raw) > 12:
                    # Split merged given names intelligently
                    # Look for capital letter patterns or split in half
                    mid = len(given_name_raw) // 2
                    given_name = given_name_raw[:mid] + ' ' + given_name_raw[mid:]
                    logger.info(f"Heuristic split applied (merged given names): original='{given_name_raw}', split='{given_name}'")
                else:
                    given_name = given_name_raw
            else:
                # OCR misread << as single <
                # Split on single < and use heuristic:
                # Typically: SURNAME<GIVENNAME1<GIVENNAME2
                # First part = surname, rest = given names
                parts = names_raw.split('<')
                # Remove empty parts
                parts = [p for p in parts if p]

                if len(parts) >= 2:
                    # First part is surname
                    surname = parts[0].strip()
                    # Rest are given names
                    given_name = " ".join(parts[1:]).strip()
                elif len(parts) == 1:
                    # Only surname, no given names
                    surname = parts[0].strip()
                    given_name = ""
                else:
                    surname = ""
                    given_name = ""

                logger.info(f"No << found, split by < into {len(parts)} parts")

            logger.info(f"Parsed Surname: '{surname}'")
            logger.info(f"Parsed Given Name: '{given_name}'")

            # Parse line 2 fields
            passport_number = line2[:9].replace("<", "")
            logger.info(f"Passport Number (pos 0-9): '{passport_number}'")

            nationality = line2[10:13].replace("<", "")
            logger.info(f"Nationality (pos 10-13): '{nationality}'")

            dob_raw = line2[13:19]  # YYMMDD
            logger.info(f"DOB raw (pos 13-19): '{dob_raw}'")

            sex = line2[20]
            logger.info(f"Sex (pos 20): '{sex}'")

            expiry_raw = line2[21:27]  # YYMMDD
            logger.info(f"Expiry raw (pos 21-27): '{expiry_raw}'")

            personal_number = line2[28:42].replace("<", "")
            logger.info(f"Personal Number (pos 28-42): '{personal_number}'")

            # Convert dates from YYMMDD to YYYY-MM-DD
            date_of_birth = self._convert_date(dob_raw)
            date_of_expiry = self._convert_date(expiry_raw)

            return {
                'passportNumber': passport_number,
                'surname': surname,
                'givenName': given_name,
                'nationality': nationality,
                'dateOfBirth': date_of_birth,
                'sex': sex if sex in ['M', 'F'] else 'M',  # Default to M if invalid
                'dateOfExpiry': date_of_expiry,
                'issuingCountry': issuing_country,
                'personalNumber': personal_number if personal_number else None,
                'rawMrz': raw_mrz
            }

        except Exception as e:
            logger.error(f"Field extraction failed: {str(e)}")
            # Return minimal data
            return {
                'passportNumber': 'UNKNOWN',
                'surname': '',
                'givenName': '',
                'nationality': '',
                'dateOfBirth': '',
                'sex': 'M',
                'dateOfExpiry': '',
                'issuingCountry': '',
                'personalNumber': None,
                'rawMrz': raw_mrz
            }

    def _convert_date(self, date_str: str) -> str:
        """
        Convert YYMMDD to YYYY-MM-DD.

        Handles Y2K: YY < 50 = 20YY, YY >= 50 = 19YY
        Example: 900315 → 1990-03-15, 250815 → 2025-08-15
        """
        if not date_str or len(date_str) != 6:
            return ""

        try:
            yy = int(date_str[0:2])
            mm = date_str[2:4]
            dd = date_str[4:6]

            # Y2K handling
            if yy >= 50:
                yyyy = 1900 + yy
            else:
                yyyy = 2000 + yy

            return f"{yyyy}-{mm}-{dd}"

        except ValueError:
            logger.warning(f"Invalid date format: {date_str}")
            return ""

    def _validate_check_digits_manual(self, mrz_text: str) -> bool:
        """
        Validate MRZ check digits using ICAO 9303 algorithm.

        MRZ check digits use weighted sum modulo 10.
        Weights: 7, 3, 1 (repeating)
        """
        try:
            def check_digit(data: str) -> int:
                """Calculate check digit for given data"""
                weights = [7, 3, 1]
                total = 0
                for i, char in enumerate(data):
                    if char == '<':
                        value = 0
                    elif char.isdigit():
                        value = int(char)
                    else:  # Letter
                        value = ord(char) - ord('A') + 10
                    total += value * weights[i % 3]
                return total % 10

            line2 = mrz_text[44:88]

            # Check digit 1: Passport number (positions 0-8, check at 9)
            passport_check = line2[9]
            if passport_check.isdigit():
                expected = check_digit(line2[:9])
                if int(passport_check) != expected:
                    logger.warning(f"Passport number check digit mismatch: {passport_check} != {expected}")
                    return False

            # Check digit 2: Date of birth (positions 13-18, check at 19)
            dob_check = line2[19]
            if dob_check.isdigit():
                expected = check_digit(line2[13:19])
                if int(dob_check) != expected:
                    logger.warning(f"DOB check digit mismatch: {dob_check} != {expected}")
                    # Don't fail, just warn
                    pass

            # Check digit 3: Expiry date (positions 21-26, check at 27)
            expiry_check = line2[27]
            if expiry_check.isdigit():
                expected = check_digit(line2[21:27])
                if int(expiry_check) != expected:
                    logger.warning(f"Expiry check digit mismatch: {expiry_check} != {expected}")
                    # Don't fail, just warn
                    pass

            return True  # Return true even if some checks fail (OCR errors common)

        except Exception as e:
            logger.error(f"Check digit validation failed: {str(e)}")
            return True  # Don't fail parsing due to check digit issues


# Singleton instance
_mrz_parser_instance: Optional[MRZParser] = None


def get_mrz_parser() -> MRZParser:
    """
    Get singleton MRZParser instance.
    """
    global _mrz_parser_instance

    if _mrz_parser_instance is None:
        _mrz_parser_instance = MRZParser()

    return _mrz_parser_instance

File: src/lib/mrzParser.js
/**
 * MRZ (Machine Readable Zone) Parser
 * Parses passport MRZ data according to ICAO Document 9303 standard
 *
 * MRZ Format (2 lines, 44 characters each = 88 total):
 * Line 1: P<ISSUINGCOUNTRYSURNAME<<GIVENNAMES<<<<<<<<<<<<<<
 * Line 2: PASSPORTNUMBER<NATIONALITY<DOBYYMMDDSEXEXPIRYYYMMDD<<<<<<<<<<<<<CHECKDIGITS
 *
 * @module mrzParser
 */

import { convertCountryCodeToNationality } from './countryCodeMapper';

/**
 * Validates if a string is a valid MRZ format
 * @param {string} input - The input string to validate
 * @returns {boolean} - True if valid MRZ format
 */
export const isMrzFormat = (input) => {
  if (!input || typeof input !== 'string') return false;

  const cleaned = input.replace(/\s/g, '');

  // Must be exactly 88 characters
  if (cleaned.length !== 88) return false;

  // Must start with P< (passport identifier)
  if (!cleaned.startsWith('P<')) return false;

  // Should contain mostly uppercase letters, numbers, and <
  const validChars = /^[A-Z0-9<]+$/;
  if (!validChars.test(cleaned)) return false;

  return true;
};

/**
 * Parse MRZ data from passport scan
 * @param {string} mrzString - The 88-character MRZ string
 * @returns {Object} Parsed passport data or error
 */
export const parseMrz = (mrzString) => {
  try {
    const cleanedMrz = mrzString.replace(/\s/g, '');

    // Validate length
    if (cleanedMrz.length < 88) {
      throw new Error(`MRZ length is too short (${cleanedMrz.length} chars, expected 88).`);
    }

    // Split into two lines
    const line1 = cleanedMrz.substring(0, 44);
    const line2 = cleanedMrz.substring(44, 88);

    // === LINE 1 PARSING ===
    // Format: P<ISSUINGCOUNTRYSURNAME<<GIVENNAMES<<<<<<<<<<<<<

    // Extract document type and issuing country
    const documentType = line1.substring(0, 2); // Should be "P<"
    const issuingCountry = line1.substring(2, 5);

    // Extract names (rest of line 1, split by <<)
    const namesSection = line1.substring(5);
    const names = namesSection.split('<<');
    const surname = names[0].replace(/</g, ' ').trim();
    const givenName = names.slice(1).join(' ').replace(/</g, ' ').trim();

    // === LINE 2 PARSING ===
    // Format: PASSPORTNUMBER<NATIONALITY<DOBYYMMDDSEXEXPIRYYYMMDD<<<<<<<<<<<<<CHECKDIGITS

    // Passport number (9 chars)
    const passportNumber = line2.substring(0, 9).replace(/</g, '').trim();
    const passportCheckDigit = line2.substring(9, 10);

    // Nationality (3 chars)
    const nationalityCode = line2.substring(10, 13);
    const nationality = convertCountryCodeToNationality(nationalityCode);

    // Date of birth (6 chars: YYMMDD)
    const dobRaw = line2.substring(13, 19);
    let dobYear = parseInt(dobRaw.substring(0, 2), 10);
    // Year conversion: if YY > current year's YY, it's 1900s, else 2000s
    const currentYear = new Date().getFullYear() % 100;
    dobYear += (dobYear > currentYear) ? 1900 : 2000;
    const dobMonth = dobRaw.substring(2, 4);
    const dobDay = dobRaw.substring(4, 6);
    const dob = `${dobYear}-${dobMonth}-${dobDay}`;
    const dobCheckDigit = line2.substring(19, 20);

    // Sex (1 char: M/F/<)
    const sexRaw = line2.substring(20, 21);
    const sex = sexRaw === 'M' ? 'Male' : sexRaw === 'F' ? 'Female' : 'Other';

    // Expiry date (6 chars: YYMMDD)
    const expiryRaw = line2.substring(21, 27);
    let expiryYear = parseInt(expiryRaw.substring(0, 2), 10);
    // Year conversion: if YY > 50, it's 1900s, else 2000s (expiry dates are usually near future)
    expiryYear += (expiryYear > 50) ? 1900 : 2000;
    const expiryMonth = expiryRaw.substring(2, 4);
    const expiryDay = expiryRaw.substring(4, 6);
    const expiryDate = `${expiryYear}-${expiryMonth}-${expiryDay}`;
    const expiryCheckDigit = line2.substring(27, 28);

    // Personal number / Optional data (14 chars)
    const personalNumber = line2.substring(28, 42).replace(/</g, '').trim();
    const personalNumberCheckDigit = line2.substring(42, 43);

    // Final composite check digit
    const compositeCheckDigit = line2.substring(43, 44);

    // Return parsed data
    return {
      success: true,
      type: 'mrz',
      raw: cleanedMrz,
      documentType: documentType,
      issuingCountry: issuingCountry,
      passportNumber: passportNumber,
      surname: surname,
      givenName: givenName,
      nationality: nationality, // Full nationality name (e.g., "Australian")
      nationalityCode: nationalityCode, // 3-letter code (e.g., "AUS")
      dob: dob,
      sex: sex,
      dateOfExpiry: expiryDate,
      personalNumber: personalNumber,
      // Include check digits for validation if needed
      checkDigits: {
        passportNumber: passportCheckDigit,
        dob: dobCheckDigit,
        expiry: expiryCheckDigit,
        personalNumber: personalNumberCheckDigit,
        composite: compositeCheckDigit
      },
      message: 'Passport MRZ parsed successfully.'
    };
  } catch (error) {
    return {
      success: false,
      type: 'error',
      message: error.message || 'Invalid MRZ format.',
      error: error
    };
  }
};

/**
 * Format passport data for display
 * @param {Object} parsedData - Parsed MRZ data
 * @returns {Object} Formatted display data
 */
export const formatMrzForDisplay = (parsedData) => {
  if (!parsedData.success) return null;

  return {
    'Passport Number': parsedData.passportNumber,
    'Full Name': `${parsedData.givenName} ${parsedData.surname}`,
    'Nationality': parsedData.nationality,
    'Date of Birth': parsedData.dob,
    'Sex': parsedData.sex,
    'Expiry Date': parsedData.dateOfExpiry,
    'Issuing Country': parsedData.issuingCountry
  };
};

/**
 * Validate MRZ check digits (optional, for future enhancement)
 * @param {string} data - The data to validate
 * @param {string} checkDigit - The check digit
 * @returns {boolean} True if valid
 */
export const validateCheckDigit = (data, checkDigit) => {
  // ICAO check digit algorithm
  const weights = [7, 3, 1];
  let sum = 0;

  for (let i = 0; i < data.length; i++) {
    const char = data[i];
    let value;

    if (char === '<') {
      value = 0;
    } else if (char >= '0' && char <= '9') {
      value = parseInt(char, 10);
    } else if (char >= 'A' && char <= 'Z') {
      value = char.charCodeAt(0) - 'A'.charCodeAt(0) + 10;
    } else {
      return false; // Invalid character
    }

    sum += value * weights[i % 3];
  }

  const calculatedDigit = (sum % 10).toString();
  return calculatedDigit === checkDigit;
};

export default {
  parseMrz,
  isMrzFormat,
  formatMrzForDisplay,
  validateCheckDigit
};
